modules/raft/src/integrationTest/java/org/apache/ignite/internal/raft/ItLearnersTest.java

```diff
@@ -0,0 +1,338 @@
+
+            assertThat(refreshMembers.thenApply(RaftGroupService::leader), willBe(follower.asPeer()));
+            assertThat(refreshMembers.thenApply(RaftGroupService::peers), will(contains(follower.asPeer())));
+            assertThat(refreshMembers.thenApply(RaftGroupService::learners), will(containsInAnyOrder(toPeerArray(learners))));
+        });
+    }
+
+    /**
+     * Tests that if the only follower is stopped, then the majority is lost.
+     */
+    @Test
+    public void testLostLeadership() throws Exception {
+        RaftNode follower = nodes.get(0);
+        List<RaftNode> learners = nodes.subList(1, nodes.size());
+
+        List<CompletableFuture<RaftGroupService>> services = nodes.stream()
+                .map(node -> startRaftGroup(node, new TestRaftGroupListener(), List.of(follower), learners))
+                .collect(toList());
+
+        // Wait for the leader to be elected.
+        services.forEach(service -> {
+            assertThat(service.thenCompose(s -> s.refreshLeader().thenApply(v -> s.leader())), willBe(follower.asPeer()));
+        });
+
+        nodes.set(0, null).close();
+
+        assertThat(services.get(1).thenCompose(s -> s.run(new TestWriteCommand("foo"))), willThrow(TimeoutException.class));
+    }
+
+    /**
+     * Tests that even if all learners are stopped, then the majority is not lost.
+     */
+    @Test
+    public void testLostLearners() throws Exception {
+        RaftNode follower = nodes.get(0);
+        List<RaftNode> learners = nodes.subList(1, nodes.size());
+
+        List<CompletableFuture<RaftGroupService>> services = nodes.stream()
+                .map(node -> startRaftGroup(node, new TestRaftGroupListener(), List.of(follower), learners))
+                .collect(toList());
+
+        // Wait for the leader to be elected.
+        // Wait for the leader to be elected.
+        services.forEach(service -> {
+            assertThat(service.thenCompose(s -> s.refreshLeader().thenApply(v -> s.leader())), willBe(follower.asPeer()));
+        });
+
+        nodes.set(1, null).close();
+        nodes.set(2, null).close();
+
+        assertThat(services.get(0).thenCompose(RaftGroupService::refreshLeader), willCompleteSuccessfully());
+    }
+
+    private CompletableFuture<RaftGroupService> startRaftGroup(
+            RaftNode raftNode,
+            RaftGroupListener listener,
+            List<RaftNode> peers,
+            List<RaftNode> learners
+    ) {
+        try {
+            CompletableFuture<RaftGroupService> future = raftNode.loza.prepareRaftGroup(
+                    RAFT_GROUP_NAME,
+                    peers.stream().map(RaftNode::localMember).collect(toList()),
+                    learners.stream().map(RaftNode::localMember).collect(toList()),
+                    () -> listener,
+                    () -> noopLsnr,
+                    RaftGroupOptions.defaults()
+            );
+
+            return future.thenApply(s -> {
+                s.timeout(100);
```

Comment:
Please add a comment for this line. I believe it is here to make tests faster

Reply:
Yes, you are right